src/binder.ts

```diff
@@ -30,173 +31,190 @@ import DapConnection from './dap/connection';
+      };
+    });
+    dap.on('configurationDone', async () => ({}));
+    dap.on('threads', async () => ({ threads: [] }));
+    dap.on('loadedSources', async () => ({ sources: [] }));
+    dap.on('breakpointLocations', () => Promise.resolve({ breakpoints: [] }));
+    dap.on('attach', async params => {
+      await this.boot(params as AnyResolvingConfiguration, dap);
+      return {};
+    });
+    dap.on('launch', async params => {
+      await this.boot(params as AnyResolvingConfiguration, dap);
+      return {};
+    });
+    dap.on('pause', async () => {
+      return {};
+    });
+    dap.on('terminate', () => this._terminateRoot());
+    dap.on('disconnect', () => this._disconnectRoot());
+    dap.on('restart', async ({ arguments: params }) => {
+      await this._restart(params as AnyResolvingConfiguration);
+      return {};
+    });
+    dap.on('startSelfProfile', async ({ file }) => {
+      selfProfile?.dispose();
+      selfProfile = new SelfProfile(file);
+      await selfProfile.start();
+      return {};
+    });
+    dap.on('stopSelfProfile', async () => {
+      if (selfProfile) {
+        await selfProfile.stop();
+        selfProfile.dispose();
+        selfProfile = undefined;
+      }
 
-        return {};
-      });
+      return {};
     });
   }
 
-  private getLaunchers() {
-    if (!this._launchers) {
-      this._launchers = new Set(this._rootServices.getAll(ILauncher));
-
-      for (const launcher of this._launchers) {
-        launcher.onTargetListChanged(
-          () => {
-            const targets = this.targetList();
-            this._attachToNewTargets(targets, launcher);
-            this._detachOrphanThreads(targets);
-            this._onTargetListChangedEmitter.fire();
-          },
-          undefined,
-          this._disposables,
-        );
-      }
+  private readonly getLaunchers = once(() => {
+    const launchers = new Set<ILauncher>(this._rootServices.getAll(ILauncher));
+
+    for (const launcher of launchers) {
+      this._disposables.push(
+        launcher.onTargetListChanged(() => {
+          const targets = this.targetList();
+          this._attachToNewTargets(targets, launcher);
+          this._terminateOrphanThreads(targets);
+        }),
+      );
     }
 
-    return this._launchers;
+    return launchers as ReadonlySet<ILauncher>;
+  });
+
+  /**
+   * Terminates all running targets. Resolves when all have terminated.
+   */
+  private async _terminateRoot() {
+    this._root.state = TargetState.Terminating;
+    await Promise.all([...this.getLaunchers()].map(l => l.terminate()));
+    await this._root.waitUntilChildrenAre(TargetState.Terminated);
+    this._root.state = TargetState.Terminated;
+    return {};
   }
 
-  private async _disconnect() {
-    if (!this._launchers) {
-      return;
+  /**
+   * Disconnects all running targets. Resolves when all have disconnected.
+   */
+  private async _disconnectRoot() {
+    if (this._root.state < TargetState.Terminating) {
```

Comment:
The "<" seems off to me. I would have expected that terminating is a "stronger state" or ">=" to disconnecting.